pkg/route/secret/secret_monitor.go

```diff
@@ -0,0 +1,192 @@
+		kubeClient: kubeClient,
+		monitors:   map[ObjectKey]*singleItemMonitor{},
+	}
+}
+
+// create secret watch.
+func (s *secretMonitor) AddSecretEventHandler(namespace, secretName string, handler cache.ResourceEventHandler) (SecretEventHandlerRegistration, error) {
+	return s.addSecretEventHandler(namespace, secretName, handler, nil)
+}
+
+// addSecretEventHandler should only be used directly for tests. For production use AddSecretEventHandler().
+// createInformerFn helps in mocking sharedInformer for unit tests.
+func (s *secretMonitor) addSecretEventHandler(namespace, secretName string, handler cache.ResourceEventHandler, createInformerFn func() cache.SharedInformer) (SecretEventHandlerRegistration, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if handler == nil {
+		return nil, fmt.Errorf("nil handler is provided")
+	}
+
+	// secret identifier (namespace/secret)
+	key := NewObjectKey(namespace, secretName)
+
+	// check if secret monitor(watch) already exists
+	m, exists := s.monitors[key]
+
+	// start secret informer
+	if !exists {
+		var sharedInformer cache.SharedInformer
+		if createInformerFn == nil {
+			// create a single secret monitor
+			sharedInformer = cache.NewSharedInformer(
+				cache.NewListWatchFromClient(
+					s.kubeClient.CoreV1().RESTClient(),
+					"secrets",
+					namespace,
+					fields.OneTermEqualSelector("metadata.name", secretName),
+				),
+				&corev1.Secret{},
+				0,
+			)
+		} else {
+			// only for testability
+			klog.Warning("creating informer for testability")
+			sharedInformer = createInformerFn()
+		}
+
+		m = newSingleItemMonitor(key, sharedInformer)
+		go m.StartInformer()
+
+		// wait for first sync
+		if !cache.WaitForCacheSync(context.Background().Done(), m.HasSynced) {
+			return nil, fmt.Errorf("failed waiting for cache sync")
+		}
+
+		// add item key to monitors map // add watch to the list
+		s.monitors[key] = m
+
+		klog.Info("secret informer started", " item key ", key)
+	}
+
+	// secret informer already started, just add the handler
+	klog.Info("secret handler added", " item key ", key)
+
+	return m.AddEventHandler(handler) // also populate key inside secretEventHandlerRegistration
+}
+
+// Remove secret watch
+func (s *secretMonitor) RemoveSecretEventHandler(handlerRegistration SecretEventHandlerRegistration) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if handlerRegistration == nil {
+		return fmt.Errorf("secret handler is nil")
+	}
+
+	// get the secret identifier linked with handler
+	// populated in AddEventHandler()
+	key := handlerRegistration.GetKey()
+
+	// check if secret informer already exists for the secret(key)
+	m, exists := s.monitors[key]
+	if !exists {
+		return fmt.Errorf("secret monitor already removed for item key %v", key)
+	}
+
+	if err := m.RemoveEventHandler(handlerRegistration); err != nil {
+		return err
+	}
+	klog.Info("secret handler removed", " item key", key)
+
+	// stop informer if there is no handler
+	if m.numHandlers.Load() <= 0 {
```

Comment:
put this counter of number of handlers into the `secretMonitor`, not the `singleItemMonitor since this is where the orchestration happens.

Reply:
Sure, added a new struct local to secret_monitor and moved `numHandlers` here. 

```go
type monitoredItem struct {
	itemMonitor *singleItemMonitor
	numHandlers atomic.Int32
}

type secretMonitor struct {
	kubeClient kubernetes.Interface
	lock       sync.RWMutex
	monitors   map[ObjectKey]*monitoredItem
}
```